### ISSUE TITLE
Marginalia - Upgrade to Plone 4; New feature requests

### ISSUE THREAD

**Issue by GoogleCodeExporter** [tamzi/bungeni-portal#740]
```
A requirement for the Bungeni-Plone system, is that it should be possible to 
annotate and amend documents through the web. This will be done via the product 
Marginalia installed on the Plone content management system

Marginalia consists of a javascript client which runs on the browser, and which 
provides the UI and rendering of annotations and amendments, and of a server 
component, which receives and stores new annotations and amendments from the 
client and sends existing annotations and amendments upon request. The 
communication between the client and the server is defined in the marginalia 
client implementation, and uses an ATOM feed for the transfer of annotations. 

Current Status
------------

An implementation of Marginalia for Plone 3 exists at: 
http://bungeni-portal.googlecode.com/svn/Marginalia/

The following key requirements are currently met by the existing Marginalia 
Product.

    * Annotatable documents are identified in Plone using the documents UID; Annotatable documents can be annotated (comments only) and amended (comments, insertions, replacements and deletions);
    * Annotatable documents are stored as Archetype-based objects in Plone, and have the normal schema, security and workflow characteristics of Archetype-based content;
    * Annotations and Amendments can reference other Plone content using the ATReferenceBrowserWidget which is made be available to the marginalia javascript client. It is possible to incorporate rich content objects within annotations by reference;
    * The UI makes available the following features and functionality:
     - filtering by user and group;
     - for amendments, filtering by amendment type;
     - private and public status of amendments and annotations (set by the user);     
     - text search filter (the filter is carried out on public amendments/annotations;
     - amendments and annotations can be deleted by the user;
     - super user view - the super user can see all public amendments/annotations. For each amendment/annotation the super user can change the state of the amendment/annotation by either accepting, declining or setting it to pending state.

Requirements
------------

1.Update Marginalia for Plone 4; 

2.Currently an annotatable document has two tabs "comment" and "amend"; Only 
one of these tabs should be available once editing of the document is complete; 
On selection of the "edit complete" flag the user will set the document as 
either "annotatable" or "amendable";

3.An annotatable document once set for annotation or amendment cannot be edited 
further; The document owner should be able to set a flag on the document to 
indicate that the editing process is complete;

4. A download feature should be provided to convert Marginalia's annotations 
and amendments into physical change markings This conversion process would 
alter the document by adding tags into the document to indicate 
annotations/amendments added by the user and accepted into the document. 
Attributes stored on the tags would include user-name i.e. the user who added 
the annotation/amendment, date and time of the change, type of 
annotation/amendment; 

5. Integration with existing Plone content types should be non-intrusive, it 
should be sufficient to apply the required marker interfaces and document 
actions via ZCML. 

6.A download feature should be provided to convert Marginalia's annotations and 
amendments into physical change markings This conversion process would alter 
the document by adding tags into the document to indicate 
annotations/amendments added by the user and accepted into the document. 
Attributes stored on the tags would include user-name i.e. the user who added 
the annotation/amendment, date and time of the change, type of 
annotation/amendment; 

7. Committing physical change markings  - An option must be provided to commit 
physical change markings into the document content. i.e. A physical change 
deletion mark must be converted to actual deletion of the text, and an 
insertion change mark must actually insert the text of the change mark into the 
document

8. Versioning - It must be possible to extract a document with physical change 
markings as a new version of the original document.

```

Original issue reported on code.google.com by `mng...@gmail.com` on 24 Nov 2011 at 11:13
- Merged into: #758
